cmake/kokkos_functions.cmake

```
@@ -2,344 +2,42 @@
-      if(NOT CMAKE_CXX_STANDARD)
-        set(INTERNAL_CXX11_STANDARD_COMPILE_OPTION ${CMAKE_CXX11_STANDARD_COMPILE_OPTION})
-        set(INTERNAL_CXX11_EXTENSION_COMPILE_OPTION ${CMAKE_CXX11_EXTENSION_COMPILE_OPTION})
-      endif()
-    endif()
-
-    # Set the C++ standard info for Kokkos respecting user set values for
-    # CMAKE_CXX_STANDARD and CMAKE_CXX_EXTENSIONS.
-    # Only use cxx extension if explicitly requested
-    if(CMAKE_CXX_STANDARD EQUAL 14)
-      if(DEFINED CMAKE_CXX_EXTENSIONS AND CMAKE_CXX_EXTENSIONS STREQUAL ON)
-        set(INTERNAL_CXX_FLAGS ${INTERNAL_CXX14_EXTENSION_COMPILE_OPTION})
-      else()
-        set(INTERNAL_CXX_FLAGS ${INTERNAL_CXX14_STANDARD_COMPILE_OPTION})
-      endif()
-    elseif(CMAKE_CXX_STANDARD EQUAL 11)
-      if(DEFINED CMAKE_CXX_EXTENSIONS AND CMAKE_CXX_EXTENSIONS STREQUAL ON)
-        set(INTERNAL_CXX_FLAGS ${INTERNAL_CXX11_EXTENSION_COMPILE_OPTION})
-      else()
-        set(INTERNAL_CXX_FLAGS ${INTERNAL_CXX11_STANDARD_COMPILE_OPTION})
-      endif()
-    else()
-      # The user didn't explicitly request a standard, transitively require
-      # C++11 respecting CMAKE_CXX_EXTENSIONS.
-      if(DEFINED CMAKE_CXX_EXTENSIONS AND CMAKE_CXX_EXTENSIONS STREQUAL ON)
-        set(INTERNAL_CXX_FLAGS ${INTERNAL_CXX11_EXTENSION_COMPILE_OPTION})
-      else()
-        set(INTERNAL_CXX_FLAGS ${INTERNAL_CXX11_STANDARD_COMPILE_OPTION})
-      endif()
-    endif()
+#   kokkos_option
+
+FUNCTION(kokkos_option CAMEL_SUFFIX DEFAULT TYPE DOCSTRING)
+  SET(CAMEL_NAME Kokkos_${CAMEL_SUFFIX})
+  STRING(TOUPPER ${CAMEL_NAME} UC_NAME)
+  SET(CACHE_NAME KOKKOS_CACHED_${UC_NAME})
+  IF (NOT DEFINED ${CACHE_NAME} AND DEFINED ${CAMEL_NAME})
+    #THIS IS our first time through the cmake
+    #WE WERE given the camel case name instead of the UC name we wanted
+    #MAKE DARn sure we don't have both an UC and Camel version that differ
+    IF (DEFINED ${UC_NAME} AND NOT ${CAMEL_NAME} STREQUAL ${UC_NAME})
+      MESSAGE(FATAL_ERROR "Given both ${CAMEL_NAME} and ${UC_NAME} with different values: ${${CAMEL_NAME}} != ${${UC_NAME}}")
+    ENDIF()
+    #GREAT, No conflicts - use the camel case name as the default for the UC
+    SET(${UC_NAME} ${${CAMEL_NAME}} CACHE ${TYPE} ${DOCSTRING})
+  ELSEIF(DEFINED ${CAMEL_NAME})
+    #THIS IS at least our second configure and we have an existing cache
+    #CMAKE Makes this impossible to distinguish something already in cache
+    #AND SOMthing given explicitly on the command line
+    #AT THIS point, we have no choice but to accept the Camel value and print a warning
+    IF (NOT ${CAMEL_NAME} STREQUAL ${UC_NAME})
+      MESSAGE(WARNING "Overriding ${UC_NAME}=${${UC_NAME}} with ${CAMEL_NAME}=${${CAMEL_NAME}}")
+    ENDIF()
+    #I HAVE to accept the Camel case value - really no choice here - force it
+    SET(${UC_NAME} ${${CAMEL_NAME}} CACHE ${TYPE} ${DOCSTRING} FORCE)
+  ELSE() #GReat, no camel case names - nice and simple
+    SET(${UC_NAME} ${DEFAULT} CACHE ${TYPE} ${DOCSTRING})
```

```suggestion
    SET(${UC_NAME} ${DEFAULT} CACHE INTERNAL ${DOCSTRING})
```

I guess you don't want this to show up in `ccmake`.

---

agreed